common/src/gas_provider/internal.rs

```diff
@@ -110,75 +144,152 @@ where
+        node: &StorageMap::Value,
+    ) -> Result<Option<(StorageMap::Value, MapKey)>, Error> {
+        match node.inner {
+            GasNodeType::External { .. } | GasNodeType::ReservedLocal { .. } => Ok(None),
+            GasNodeType::SpecifiedLocal { parent, .. } => {
+                let mut ret_id = parent;
+                let mut ret_node =
+                    Self::get_node(parent).ok_or_else(InternalError::parent_is_lost)?;
+                while !ret_node.is_patron() {
+                    match ret_node.inner {
+                        GasNodeType::External { .. } => return Ok(None),
+                        GasNodeType::SpecifiedLocal { parent, .. } => {
+                            ret_id = parent;
+                            ret_node =
+                                Self::get_node(parent).ok_or_else(InternalError::parent_is_lost)?;
+                        }
+                        _ => return Err(InternalError::unexpected_node_type().into()),
+                    }
+                }
+                Ok(Some((ret_node, ret_id)))
+            }
+            // Although unspecified local type has a patron parent, it's considered
+            // an error to call the method from that type of gas node.
+            GasNodeType::UnspecifiedLocal { .. } => Err(InternalError::forbidden().into()),
+        }
+    }
+
+    /// Tries to remove consumed nodes on the same path from the `key` node to the
+    /// root (including it). While trying to remove nodes, also catches value stored
+    /// in them is performed.
+    ///
+    /// Value catch is performed for all the non-patron nodes on the path from `key` to root,
+    /// until some patron node is reached. By the invariant, catching can't be blocked,
+    /// because the node is not a patron.
+    ///
+    /// For node removal there are 2 main requirements:
+    /// 1. it's not a patron node
+    /// 2. it doesn't have any children nodes.
+    ///
+    /// Although the value in nodes is moved or returned to the origin, calling `ValueNode::catch_value`
+    /// in this procedure can still result in catching non-zero value. That's possible for example, when
+    /// Gas-ful parent is consumed and has a gas-less child. When gas-less child is consumed in `ValueTree::consume`
+    /// call, the gas-ful parent's value is caught in this function.
+    ///
+    /// # Invariants
+    /// Internal invariant of the procedure:
+    /// 1. If `catch_value` call ended up with `CatchValueOutput::Missed` in `consume`, all the calls of catch_value on ancestor nodes will be `CatchValueOutput::Missed` as well.
+    /// 2. Also in that case cascade ancestors consumption will last until either the patron node or the first ancestor with specified child is found.
+    /// 3. If `catch_value` call ended up with `CatchValueOutput::Caught(x)` in `consume`, all the calls of `catch_value` on ancestor nodes will be `CatchValueOutput::Caught(0)`.
+    pub(super) fn try_remove_consumed_ancestors(
```

Comment:
This function has a quadratic complexity with respect to the length of the path from a node to it's patron (in the worst case to the root node): there is a while loop, and another nested loop inside the `catch_value()` function. The purpose of that seems to be to check whether a consumed ancestor still has a patron node somewhere upstream or not, if I'm getting it right.
This function is called every time a node is consumed so it can potentially introduce some overhead for certain tree configurations. How bad could it be or can we do better than O(n^2)?

Reply:
Discussed, will be optimised in the next PR

Reply:
> Discussed, will be optimised in the next PR

Please, always open issues for such things